fasjson_client/client.py

```diff
@@ -0,0 +1,118 @@
+import errno
+from urllib.parse import urlparse
+
+import yaml
+import requests
+from requests_gssapi import HTTPSPNEGOAuth
+from bravado import requests_client
+from bravado.client import SwaggerClient
+import gssapi
+from swagger_spec_validator.common import SwaggerValidationError
+
+from . import errors
+
+
+class HttpClient(requests_client.RequestsClient):
+  """
+  HttpClient class used in SwaggerClient for gssapi authentication.
+  """
+
+  def __init__(self, principal, *args, **kwargs):
+      super().__init__(*args, **kwargs)
+      self.principal = principal
+
+  def authenticated_request(self, request_params):
+    """
+    Retrieves an authentication token from kerberos
+    and adds it in the http header request.
+    """
+    name = gssapi.Name(self.principal, gssapi.NameType.kerberos_principal)
+    try:
+      creds = gssapi.Credentials(name=name, usage="initiate")
+    except gssapi.raw.misc.GSSError as e:
+      data = {
+        "trace": repr(e.gen_message),
+        "codes": {
+          "maj": e.maj_code,
+          "min": e.min_code,
+          "routine": e.routine_code,
+          "supplementary_code": e.supplementary_code,
+        },
+      }
+      raise errors.ClientError(
+        "schema validation failed", errno.EPROTO, data=data
+      )
+
+    request_params["auth"] = HTTPSPNEGOAuth(creds=creds)
+
+    return super().authenticated_request(request_params)
+
+
+class Client:
+  """
+  FasJsonClient client class that builds API methods based on openapi specs.
+  """
+
+  def __init__(self, spec, principal):
+    self.principal = principal
+    self.http_client = HttpClient(principal)
+    try:
+      self._api = SwaggerClient.from_spec(
+        spec,
+        http_client=self.http_client,
+        config={"disable_fallback_results": True},
+      )
+    except SwaggerValidationError:
+      raise errors.ClientError("schema validation failed", errno.EPROTO)
```

Comment:
It would be nice to have a reference to the original exception somewhere, and maybe even the relevant fields from that exception.

Reply:
That comment is still valid I think.

Reply:
just pushed a change for it.